### PR TITLE
Adding carousel support to tab controller

### PR DESCRIPTION
**Pull request by ethanshar**
In general, I added the option for the user to wrap `TabPage`s with `PageCarousel` component.
this will change the behavior of the TabController to render the pages as carousel instead of single page each time.

I made some changes in `Carousel` component to allow it to accept `Animated.event` for `onScroll` instead of a function. This way I can intercept the carousel offset position to use in the TabController.

### CODE REVIEW

**Comment by Inbal-Tish** [comment]
@ethanshar I have general questions about the design: 
First, In `PageCarousel` component, why use `Carousel` instead of `Animated.ScrollView` directly? I see that the `TabController` already has all the features for selecting index and callback for index change, so what is the `Carousel` benefit here? 
Second, why pass every `TabPage` the `asCarouselPage` prop instead of having something like `useCarousel` option on the `TabController` that, when true, renders the children (the `TabPage`s) in the `PageCarousel` component. 
Third, why expose the `PageCarousel` component as a container and not use it internally from `TabController`? Or moreover, why to have the `TabPage`s in a container and not use something like `FlatList`'s `renderItem` function?

**Comment by ethanshar** [comment]
@Inbal-Tish 
> First, In PageCarousel component, why use Carousel instead of Animated.ScrollView directly? I see that the TabController already has all the features for selecting index and callback for index change, so what is the Carousel benefit here?

I tried using `Animated.ScrollView` directly, but the layout didn't work well so I thought I might need to duplicate the code that we do for children, with wrapping the children. But I agree, it should have been enough. I'll try again. I also prefer avoiding changing things in Carousel.


> Second, why pass every TabPage the asCarouselPage prop instead of having something like useCarousel option on the TabController that, when true, renders the children (the TabPages) in the PageCarousel component.

Agree, that was the main intention, I'll fix it.

> Third, why expose the PageCarousel component as a container and not use it internally from TabController? Or moreover, why to have the TabPages in a container and not use something like FlatList's renderItem function?

I don't believe it's impossible. The current solution allows the users to render the TabPage anywhere in the hierarchy. I can't control it. and also can't wrap the pages for them. it work similar to Carousel, when you have to render elements inside it.

File: demo/src/screens/incubatorScreens/TabControllerScreen/index.js [hunk 1]
```diff
@@ -6,18 +6,19 @@ import Tab1 from './tab1';
...
+const TABS = ['tab1', 'tab2', 'tab3', 'account', 'groups', 'blog'];
 
 class TabControllerScreen extends Component {
   state = {
     selectedIndex: 0,
     items: [
       ..._.map(TABS, tab => ({label: tab, key: tab})),
-      {key: 'addTabs', icon: Assets.icons.settings, ignore: true, onPress: this.addTab},
+      {key: 'addTabs', icon: Assets.icons.settings, ignore: true, onPress: this.addTab}
     ],
```
Comment (Inbal-Tish) [inline demo/src/screens/incubatorScreens/TabControllerScreen/index.js]: extra line

File: src/incubator/TabController/index.js [hunk 2]
```diff
@@ -71,18 +80,31 @@ class TabController extends Component {
 
   render() {
     const {itemStates, ignoredItems} = this.state;
+    
+    // Rounding on Android, cause it cause issues when comparing values
+    const screenWidth = Constants.isAndroid ? Math.round(Constants.screenWidth) : Constants.screenWidth;
```
Comment (Inbal-Tish) [inline src/incubator/TabController/index.js]: Is landscape support is on the todo list?

Comment (ethanshar) [reply src/incubator/TabController/index.js]: yes, it won't work on Landscape at the moment I'm aware.

File: demo/src/screens/incubatorScreens/TabControllerScreen/index.js (hunk 1 above)
Comment (ethanshar) [reply demo/src/screens/incubatorScreens/TabControllerScreen/index.js]: extra line where?

Comment (Inbal-Tish) [reply demo/src/screens/incubatorScreens/TabControllerScreen/index.js]: line 19. Nevermind.

**Pull request merged**